File: app/src/main/kotlin/bumblebee/core/subscription/Subscription.kt
package bumblebee.core.subscription

import cn.hutool.core.date.DatePattern
import cn.hutool.core.date.DateUtil
import com.alibaba.fastjson.JSON
import io.netty.handler.codec.mqtt.MqttQoS
import java.io.Serializable
import java.util.*


class Subscription(
    val clientId: String,
    val topic: String,
    val qos: MqttQoS = MqttQoS.AT_LEAST_ONCE,
    private val createTimeStr: String = DateUtil.format(
        Date(), DatePattern.PURE_DATETIME_PATTERN
    )
) : Serializable {
    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other == null || javaClass != other.javaClass) {
            return false
        }
        val that = other as Subscription
        return clientId == that.clientId && topic == that.topic
    }

    override fun hashCode(): Int {
        return Objects.hash(clientId, topic)
    }

    override fun toString(): String {
        return JSON.toJSONString(this)
    }
}

File: src/clojure/bumblebee/mqtt/store/subscription.clj
; TODO: Move to store namespace
(ns bumblebee.mqtt.store.subscription
  (:import [io.netty.handler.codec.mqtt MqttQoS]))

(defrecord Subscription
  [client-id topic qos create-time-str])

(defn make-subscription
  "Create an idiomatic Clojure subscription value.
  Defaults QoS to AT_LEAST_ONCE and create-time-str to empty."
  ([client-id topic]
   (->Subscription client-id topic MqttQoS/AT_LEAST_ONCE ""))
  ([client-id topic qos]
   (->Subscription client-id topic qos ""))
  ([client-id topic qos create-time-str]
   (->Subscription client-id topic qos create-time-str)))
